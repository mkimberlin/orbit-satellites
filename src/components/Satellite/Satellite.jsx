import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import './styles.module.css';

const displayName = 'Satellite';
const propTypes = {
  color: PropTypes.shape({
    r: PropTypes.number,
    g: PropTypes.number,
    b: PropTypes.number,
    a: PropTypes.number
  }),
  orbit: PropTypes.number,
  size: PropTypes.oneOf([1, 2, 3])
};

const Satellite = ({ color, orbit = 1, size = 1 }) => {
  const styles = {
    backgroundColor: color
      ? `rgba(${color.r}, ${color.g}, ${color.b}, ${color.a})`
      : undefined,
    boxShadow: color
      ? `0 0 1rem rgba(${color.r}, ${color.g}, ${color.b}, ${color.a})`
      : undefined
  };

  return (
    <div
      style={styles}
      className={cx('satellite', `satellite-${orbit}-${size}`)}
    />
  );
};

Satellite.displayName = displayName;
Satellite.propTypes = propTypes;

export default Satellite;
