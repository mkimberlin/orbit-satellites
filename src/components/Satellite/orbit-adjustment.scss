@import '../../styles/animations.scss';
@import '../../styles/common-variables.scss';

@mixin orbit-adjustment(
  $orbit,
  $satellite-size,
  $tilt-x: $tilt-x,
  $tilt-y: $tilt-y,
  $orbit-base: $orbit-base,
  $orbit-step: $orbit-step,
  $satellite-step: $satellite-step
) {
  $orbit-size: $orbit-base + ($orbit * $orbit-step);
  $satellite-computed-size: $satellite-size * $satellite-step;
  $satellite-shift: ($orbit-size / 2) - ($satellite-computed-size / 2);

  @include keyframe(invert-orbit-#{$orbit}-#{$satellite-size}) {
    0% {
      transform: translate(#{$satellite-shift}) rotateZ(360deg)
        rotateY(calc(#{$tilt-y} * -1)) rotateX(calc(#{$tilt-x} * -1));
    }

    100% {
      transform: translate(#{$satellite-shift}) rotateZ(0deg)
        rotateY(calc(#{$tilt-y} * -1)) rotateX(calc(#{$tilt-x} * -1));
    }
  }
}

@mixin generate-orbit-adjustments(
  $tilt-x: $tilt-x,
  $tilt-y: $tilt-y,
  $max-orbits: $max-orbits,
  $satellite-sizes: $satellite-sizes
) {
  @for $orbit from 1 through $max-orbits {
    @for $satellite-size from 1 through $satellite-sizes {
      @include orbit-adjustment($orbit, $satellite-size, $tilt-x, $tilt-y);
    }
  }
}

@include generate-orbit-adjustments();
