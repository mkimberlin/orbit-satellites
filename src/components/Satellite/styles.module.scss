@import './orbit-adjustment.scss';

.satellite {
  background-color: rgba($default-satellite-color, 0.8);
  box-shadow: 0 0 1rem rgba($default-satellite-color, 0.8);
  border-radius: 50%;
  text-align: center;
  position: absolute;
  top: 50%;
  left: 50%;
  transform-style: preserve-3d;
  overflow: hidden;
  text-overflow: ellipsis;

  @for $orbit from 1 through $max-orbits {
    @for $satellite-size from 1 through $object-sizes {
      &.satellite-#{$orbit}-#{$satellite-size} {
        width: $satellite-size * $satellite-step;
        height: $satellite-size * $satellite-step;
        line-height: $satellite-size * $satellite-step;
        @include animation(
          calc(#{$speed-base} + (#{$speed-step} * #{$orbit - 1})),
          invert-orbit-#{$orbit}-#{$satellite-size}
        );
      }
    }
  }
}
