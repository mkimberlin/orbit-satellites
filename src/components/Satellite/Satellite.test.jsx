import React from 'react';
import Satellite from './Satellite';

describe('Satellite component', () => {
  it('renders correctly with default props', () => {
    snapshot(<Satellite />);
  });

  it('renders correctly with alternate props', () => {
    snapshot(
      <Satellite
        color={{ r: 255, g: 205, b: 155, a: 0.5 }}
        orbit={2}
        size={3}
      />
    );
  });
});
