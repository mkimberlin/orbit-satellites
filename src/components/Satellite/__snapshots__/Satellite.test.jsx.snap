// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Satellite component renders correctly with alternate props 1`] = `
<div
  className="satellite satellite-2-3"
  style={
    Object {
      "backgroundColor": "rgba(255, 205, 155, 0.5)",
      "boxShadow": "0 0 1rem rgba(255, 205, 155, 0.5)",
    }
  }
/>
`;

exports[`Satellite component renders correctly with default props 1`] = `
<div
  className="satellite satellite-1-1"
  style={
    Object {
      "backgroundColor": undefined,
      "boxShadow": undefined,
    }
  }
/>
`;
